From 7b498e6db2539e654ba663ca44b22ccb96442c55 Mon Sep 17 00:00:00 2001
From: "guoyin.chen" <guoyin.chen@freescale.com>
Date: Mon, 5 Jan 2015 10:27:24 +0800
Subject: [PATCH 558/635] MA-6070 no change when adjust the brightness from
 the middle to high

Map brightness level from framework to brightness driver's level
Make sure any brightness level greater than 0 have a minimal brightness
level in driver.

Signed-off-by: guoyin.chen <guoyin.chen@freescale.com>
---
 lights/fsl_lights.c | 29 ++++++++++++-----------------
 1 file changed, 12 insertions(+), 17 deletions(-)

diff --git a/lights/fsl_lights.c b/lights/fsl_lights.c
index 3edc44d..dde1e8f 100644
--- a/lights/fsl_lights.c
+++ b/lights/fsl_lights.c
@@ -1,6 +1,6 @@
 /*
  * Copyright (C) 2008 The Android Open Source Project
- * Copyright 2009-2014 Freescale Semiconductor, Inc.
+ * Copyright 2009-2015 Freescale Semiconductor, Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -63,8 +63,6 @@ static int set_light_backlight(struct light_device_t* dev,
     int result = -1;
     unsigned int color = state->color;
     unsigned int brightness = 0, max_brightness = 0;
-    unsigned int tmp_br = 0;
-    unsigned char tmp = 0x80;
     unsigned int i = 0;
     FILE *file;
 
@@ -81,28 +79,25 @@ static int set_light_backlight(struct light_device_t* dev,
     fclose(file);
 
     max_brightness = atoi((char *) &max_brightness);
-    //brightness = brightness * max_brightness / MAX_BRIGHTNESS;
-    for (i = 0; i < 8; i++)
-    {
-        if (tmp & brightness) {
-            tmp_br = 7 - i;
-            break;
-        } else {
-            tmp = tmp >> 1;
-        }
-    }
-    if (tmp_br > max_brightness) {
-        tmp_br = max_brightness;
+    /* any brightness greater than 0, should have at least backlight on */
+    if (max_brightness < MAX_BRIGHTNESS)
+        brightness = max_brightness *(brightness + MAX_BRIGHTNESS / max_brightness - 1) / MAX_BRIGHTNESS;
+    else
+        brightness = max_brightness * brightness / MAX_BRIGHTNESS;
+
+    if (brightness > max_brightness) {
+        brightness  = max_brightness;
     }
+    
     ALOGV("set_light, max_brightness=%d, target brightness=%d",
-        max_brightness, tmp_br);
+        max_brightness, brightness);
 
     file = fopen(path, "w");
     if (!file) {
         ALOGE("can not open file %s\n", path);
         return result;
     }
-    fprintf(file, "%d", tmp_br);
+    fprintf(file, "%d", brightness);
     fclose(file);
 
     result = 0;
-- 
1.8.0

